tests/test_source.py

```diff
@@ -0,0 +1,119 @@
+# pylint: disable=redefined-outer-name
+import typing as tp
+
+import attr
+import mock
+import plumbum as pb
+import pytest
+
+from benchbuild import source
+from benchbuild.source import BaseSource, Variant
+
+Variants = tp.Iterable[Variant]
+
+
+@attr.s
+class SimpleSource(BaseSource):
+    test_versions: tp.List[str] = attr.ib()
+
+    @property
+    def default(self) -> Variant:
+        return Variant(owner=self, version=self.test_versions[0])
+
+    def version(self, target_dir: str, version: str) -> pb.LocalPath:
+        return pb.local.path('.') / f'bb-test-{version}'
+
+    def versions(self) -> Variants:
+        return [Variant(self, v) for v in self.test_versions]
+
+
+@pytest.fixture
+def versions_a():
+    return ["1.0a", "2.0a"]
+
+
+@pytest.fixture
+def versions_b():
+    return ["1.0b", "2.0b"]
+
+
+@pytest.fixture
+def src_a(versions_a):
+    return SimpleSource(local='src_A_local',
+                        remote='src_A_remote',
+                        test_versions=versions_a)
+
+
+@pytest.fixture
+def src_b(versions_b):
+    return SimpleSource(local='src_B_local',
+                        remote='src_B_remote',
+                        test_versions=versions_b)
+
+
+def test_base_context(src_a):
+    var = src_a.default
+    ctx = source.context(var)
+
+    assert ctx[src_a.local].owner == src_a
+    assert ctx[src_a.local] == var
+
+    with pytest.raises(KeyError):
+        ctx['non-existing']  # pylint: disable=pointless-statement
+
+
+def test_base_to_str(src_a, src_b, versions_a, versions_b):
+    vars_a = src_a.versions()
+    vars_b = src_b.versions()
+
+    version_text = source.to_str(*vars_a)
+    assert version_text == ",".join(versions_a)
+    version_text = source.to_str(*vars_b)
+    assert version_text == ",".join(versions_b)
+
+    vars_ab = versions_a + versions_b
+    version_text = source.to_str(*vars_a, *vars_b)
+    assert version_text == ",".join(vars_ab)
+
+
+def test_base_default(src_a, src_b):
+    default_versions = source.default(src_a, src_b)
+
+    assert src_a.local in default_versions
+    assert default_versions[src_a.local] == src_a.default
+    assert src_b.local in default_versions
+    assert default_versions[src_b.local] == src_b.default
```

Comment:
Codacy found an issue: [Use of assert detected. The enclosed code will be removed when compiling to optimised byte code.](https://app.codacy.com/manual/polyjit/benchbuild_2/pullRequest?prid=5694346)